### ISSUE TITLE
Inconsistent handling of rvalue reference lifetimes

### ISSUE THREAD

**Issue by U007D** [rust-lang/rust#41187]
A thread that began with "can you explain this compiler error message" has uncovered inconsistent behavior in the Rust compiler.  [Here](https://users.rust-lang.org/t/rvalue-temporaries-references-their-lifetimes/10299/9) is the thread.

TL;DR:
```rust
fn main() {
    let foo = 42;
    let bar = &foo;
    assert!(bar == &42);
}
```
compiles but
```rust
fn main() {
    let foo = 42;
    let bar = Some(&foo);

    assert!(bar == Some(&42));
}
```
doesn't.

Interestingly, replacing the above non-compiling assert with:
```rust
assert!(Some(&foo) == Some(&42));
```
*does* compile.  Note `Some(&foo)` is what was assigned to `bar`, which didn't compile.

Further, replacing the original, non-working:
```rust
assert!(bar == Some(&42));
```
with:
```rust
assert_eq!(bar, Some(&42));
```
also compiles.

This inconsistent behavior is not limited to the `Option` type.  The same behavior is observed passing rvalue references to static methods and free functions.

Possible root cause uncovered by @krdln: `PartialEq()` implementations which require both their *lhs* and *rhs* operands to be of exactly the same type (including lifetimes).  This is not the case for all implementations, but it for `Option`'s `PartialEq()`, for example, it is.  More [here](https://users.rust-lang.org/t/rvalue-temporaries-references-their-lifetimes/10299/5?u=u_007d).

**Comment by ashtneoi** [comment]
It looks like this specific issue was fixed in 1.21.0 (based on a manual bisect).

**Comment by U007D** [comment]
Thank you for the heads up, @ashtneoi.  It does appear to be working now.
